### PR TITLE
Fix CORS preflight requests

### PR DESCRIPTION
**Pull request by masad-frost**
Currently the custom CORS middleware that the server has doesn't handle `OPTIONS` requests which are used to check the CORS policy of the server.

Got rid of the custom cors middleware and express's official `cors` middleware which handles all the nitty gritty of CORS requests. I made sure I pass all current CORS header values to the `cors` middleware

We use this API at repl.it for people to search for packages and add them to their HTML. All our requests are preflighted making the API unusable for us. We circumvented the issue by removing the content type (json) header, and will probably proxy the API through our servers for extra safety until this is deployed.

Thank you!

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -48,17 +49,20 @@ module.exports = () => {
...
-        res.header('Access-Control-Allow-Origin', '*');
-        res.header('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE,OPTIONS');
-        res.header('Access-Control-Allow-Headers', 'X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version');
-        if (!localMode) {
+    // Set up cors headers
+    app.use(cors({
+        origin: '*',
+        credentials: true,
+        methods: ['GET', 'PUT', 'POST', 'DELETE', 'OPTIONS'],
+        allowedHeaders: ['X-CSRF-Token', ' X-Requested-With', ' Accept', ' Accept-Version', ' Content-Length', ' Content-MD5', ' Content-Type', ' Date', ' X-Api-Version'],
```
Comment (MattIPv4) [inline app.js:57]: There seems to be some extra spacing inside the strings here, is that intended?

**Comment by MattIPv4** [comment]
I wonder if it'd also be worth adding options tests to each route, just to ensure this isn't accidentally broken down the road?

File: app.js (hunk 1 above)
Comment (masad-frost) [reply app.js:57]: Oh my bad! I'll add tests for this.

**Comment by masad-frost** [comment]
Yeah I can add exhaustive tests!

**Comment by MattIPv4** [comment]
Hey @masad-frost, just wanted to check in on this. If you don't have time to add these extras, I'm happy to either push to your fork, or to merge this and then make the changes after. Let me know what works best for you! :)

**Comment by masad-frost** [comment]
Hey, sorry about this. I can do it this weekend. Feel free to merge and I'll follow up

File: tests/suite/index.js [hunk 2]
```diff
@@ -16,6 +16,13 @@ describe('/', () => {
         expect(response).to.have.header('Cache-Control', 'public, max-age=30672000, immutable'); // 355 days
         done();
     });
+    it('returns the correct CORS and Cache headers for OPTIONS', done => {
+        request().options('/').redirects(0).end((err, res) => {
+            expect(res).to.have.header('Access-Control-Allow-Origin', '*');
+            expect(res).to.have.header('Cache-Control', 'public, max-age=30672000, immutable'); // 355 days
+            done();
+        });
+    });
```
Comment (MattIPv4) [inline tests/suite/index.js:25]:
```suggestion
    it('returns the correct CORS headers for OPTIONS', done => {
        request().options('/').redirects(0).end((err, res) => {
            expect(res).to.have.header('Access-Control-Allow-Origin', '*');
            done();
        });
    });
```

This currently fails when tests are run, I'm guessing custom headers aren't mirrored from the GET routes, which is fine.

Comment (masad-frost) [reply tests/suite/index.js:25]: Oh yeah that makes sense. We can add it, but we `Cache-Control` doesn't really make sense for `options`

Comment (drewfreyling) [reply tests/suite/index.js]: agreed

**Comment by masad-frost** [comment]
Updated, added to tests to all routes. Disclaimer, I used macros to add those since all the tests have the same format! So hopefully my laziness didn't make me miss anything.

File: tests/cors.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+const { it } = require('mocha');
+const { expect } = require('chai');
+const request = require('./base');
+const { corsOptions } = require('../utils/cors');
+
+const expectCORSHeaders = (response) => {
```
Comment (MattIPv4) [inline tests/cors.js:6]:
```suggestion
const expectCORSHeaders = response => {
```
Sorry to be pedantic, but I like to have consistency :)

File: tests/cors.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    expect(response).to.have.header('Access-Control-Allow-Origin', corsOptions.origin);
+    expect(response).to.have.header('Access-Control-Allow-Credentials', corsOptions.credentials.toString());
+
+    // testing allowed methods and headers only makes sense in case of options
```
Comment (MattIPv4) [inline tests/cors.js:10]:
```suggestion
    // Testing allowed methods and headers only makes sense in case of options
```

File: tests/cors.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+const testCors = (path, getResponse) => {
+    it('returns the correct CORS headers', done => {
+        const response = getResponse();
+        expectCORSHeaders(response);
+        done();
+    });
+
+    it('returns the correct CORS headers for OPTIONS request', (done) => {
```
Comment (MattIPv4) [inline tests/cors.js:24]:
```suggestion
    it('returns the correct CORS headers for OPTIONS request', done => {
```

File: tests/suite/libraries.js [hunk 6]
```diff
@@ -193,16 +198,16 @@ describe('/libraries', function () {
...
             let response;
             before('fetch endpoint', done => {
                 test().end((err, res) => {
                     response = res;
                     done();
                 });
             });
-            it('returns the correct CORS and Cache headers', done => {
-                expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+            it('returns the correct Cache headers', done => {
```
Comment (MattIPv4) [inline tests/suite/libraries.js:210]: CORS test missing

File: tests/suite/libraries.js [hunk 7]
```diff
@@ -240,16 +245,16 @@ describe('/libraries', function () {
...
```
Comment (MattIPv4) [inline tests/suite/libraries.js:257]: CORS test missing

File: tests/suite/libraries.js [hunk 8]
```diff
@@ -276,16 +281,16 @@ describe('/libraries', function () {
...
```
Comment (MattIPv4) [inline tests/suite/libraries.js:293]: CORS test missing

File: tests/suite/libraries.js [hunk 9]
```diff
@@ -327,16 +332,16 @@ describe('/libraries', function () {
...
```
Comment (MattIPv4) [inline tests/suite/libraries.js:344]: CORS test missing

File: tests/suite/library.js [hunk 10]
```diff
@@ -136,16 +141,16 @@ describe('/libraries/:library/:version', () => {
...
             let response;
             before('fetch endpoint', done => {
                 test().end((err, res) => {
                     response = res;
                     done();
                 });
             });
-            it('returns the correct CORS and Cache headers', done => {
-                expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+            it('returns the correct Cache headers', done => {
```
Comment (MattIPv4) [inline tests/suite/library.js:153]: CORS test missing

File: tests/suite/library.js [hunk 11]
```diff
@@ -164,16 +169,16 @@ describe('/libraries/:library/:version', () => {
...
```
Comment (MattIPv4) [inline tests/suite/library.js:181]: CORS test missing

File: tests/suite/library.js [hunk 12]
```diff
@@ -247,16 +252,16 @@ describe('/libraries/:library', () => {
...
```
Comment (MattIPv4) [inline tests/suite/library.js:264]: CORS test missing

File: tests/suite/library.js [hunk 13]
```diff
@@ -278,16 +283,16 @@ describe('/libraries/:library', () => {
...
```
Comment (MattIPv4) [inline tests/suite/library.js:295]: CORS test missing

File: tests/suite/library.js [hunk 14]
```diff
@@ -322,16 +327,16 @@ describe('/libraries/:library', () => {
...
```
Comment (MattIPv4) [inline tests/suite/library.js:339]: CORS test missing

File: tests/suite/tutorials.js [hunk 15]
```diff
@@ -148,16 +153,16 @@ describe('/libraries/:library/tutorials/:tutorial', () => {
...
                 let response;
                 before('fetch endpoint', done => {
                     test().end((err, res) => {
                         response = res;
                         done();
                     });
                 });
-                it('returns the correct CORS and Cache headers', done => {
-                    expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+                it('returns the correct Cache headers', done => {
```
Comment (MattIPv4) [inline tests/suite/tutorials.js:165]: CORS test missing

File: tests/suite/tutorials.js [hunk 16]
```diff
@@ -180,16 +185,16 @@ describe('/libraries/:library/tutorials/:tutorial', () => {
...
```
Comment (MattIPv4) [inline tests/suite/tutorials.js:197]: CORS test missing

File: tests/suite/tutorials.js [hunk 17]
```diff
@@ -211,16 +216,16 @@ describe('/libraries/:library/tutorials/:tutorial', () => {
...
```
Comment (MattIPv4) [inline tests/suite/tutorials.js:228]: CORS test missing

File: tests/suite/tutorials.js [hunk 18]
```diff
@@ -246,16 +251,16 @@ describe('/libraries/:library/tutorials/:tutorial', () => {
...
```
Comment (MattIPv4) [inline tests/suite/tutorials.js:263]: CORS test missing

File: tests/suite/tutorials.js [hunk 19]
```diff
@@ -274,16 +279,16 @@ describe('/libraries/:library/tutorials/:tutorial', () => {
...
```
Comment (MattIPv4) [inline tests/suite/tutorials.js:291]: CORS test missing

File: tests/suite/update.js [hunk 20]
```diff
@@ -1,9 +1,11 @@
 const { describe, it, before } = require('mocha');
 const { expect } = require('chai');
 const request = require('../base');
+const testCors = require('../cors');
 
 // This endpoint is a prod-only endpoint for monitoring the API auto-update job by maintainers
 describe('/update', () => {
+    const path = '/update';
     const test = () => request().get('/update');
```
Comment (MattIPv4) [inline tests/suite/update.js:9]:
```suggestion
    const test = () => request().get(path);
```

File: tests/cors.js (hunk 3 above)
Comment (masad-frost) [reply tests/cors.js:6]: Seems like something `--fix` should fix.

**Comment by MattIPv4** [comment]
Hey @masad-frost,

Would you like me to handle rebasing this and adding in the missing CORS tests, so that we can get this merged?

**Comment by masad-frost** [comment]
On my list today! Thanks for the reminder

**Comment by MattIPv4** [comment]
Hey @masad-frost -- apologies for pinging again but I just wanted to check in on this as it sounds this was originally created due to this causing issues for you in production, so we'd love to get this shipped to help y'all out.

If you don't think you'll have the time to get this ready to be landed I'm happy to take over and add the missing tests + rebasing :)

**Comment by masad-frost** [comment]
Oh boy, my bad. On it!

**Comment by masad-frost** [comment]
Should be good to go now!

**Comment by masad-frost** [comment]
Thanks for pulling this in and maintaining cdnjs!

No rush, we're proxying requests through our domain right now. Here it is in action:
cdnjscool

**Comment by MattIPv4** [comment]
Heh, awesome! Love seeing easy to use UI for bring cdnjs libraries in.

My one suggestion would be to encourage you to default to including the integrity attribute on the asset includes (CSS & JS) — SRI data should be available in the library API responses :)

**Comment by masad-frost** [comment]
Sweet, will add it.

**Pull request merged**